### PR TITLE
Introduce named types

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

Introduces a header providing named types for common internals.
The current use-case is to make types and locations more readable.

Related to #248 #1045

## Author's checklist

* [ ] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [ ] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [ ] I sticked to C++14 features.
* [ ] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?
* [ ] (more questions/tasks)

### CODE REVIEW

**Comment by DavidSCN** [comment]
Very nice, thanks! I will later go through the files more in detail. The most obvious question (and difference towards my suggestions): why not using `unsigned int` here or asking in a different way: what is the purpose of having -2e9 IDs available which we can never use? For meshIDs etc. this doesn't really matter, but for vertices, it should. The change would even be API compatible since we extend the current ID range.

In a follow up, I would like to introduce then some common constants referring to this types, e.g. `static const DataID invalidDataID = -1` or so in order to simplify initialization.

**Comment by fsimonis** [comment]
> The most obvious question (and difference towards my suggestions): why not using unsigned int here or asking in a different way: what is the purpose of having -2e9 IDs available which we can never use?

This PR serves readability/maintainability only and doesn't change types.

For the interesting discussion regarding the types itself, please continue in #887.

File: src/precice/types.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * Type used for the IDs of data
+ */
+using DataID = int;
+
+/**
+ * Type used for the IDs of meshes
+ */
+using MeshID = int;
+
+using Rank = int;
```
Comment (DavidSCN) [inline src/precice/types.hpp:35]: Some documentation (though it seems obvious) might be helpful

File: src/precice/types.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Type used for the IDs of triangles
+ */
+using TriangleID = int;
+
+/**
+ * Type used for the IDs of matching entities
+ */
+using MatchID = int;
```
Comment (DavidSCN) [inline src/precice/types.hpp:23]: In my PR, @uekerman requested some documentation clarification on this.

**Review by DavidSCN — COMMENTED** [review]
Overall very nice. I'm not sure if we can already spot every occurrence of the readable datatypes. For some of the data types, I would expect that I can change the alias and the code still compiles (e.g. for VertexID). However, this is not the case at the moment and I'm not sure if that's the purpose of this PR (feel free to try it out). Maybe, we  can also clean this up in a follow-up PR.

File: src/com/Communication.cpp [hunk 3]
```diff
@@ -53,7 +55,7 @@ void Communication::reduceSum(double const *itemsToSend, double *itemsToReceive,
   }
 }
 
-void Communication::reduceSum(double const *itemsToSend, double *itemsToReceive, int size, int rankMaster)
+void Communication::reduceSum(double const *itemsToSend, double *itemsToReceive, int size, Rank rankMaster)
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/com/Communication.cpp:58]: Good point! The solution here would be to use a span. I had my eyes on a drop-in implementation for quite some time now.

That's for another PR though.

File: src/acceleration/impl/ParallelMatrixOperations.hpp [hunk 4]
```diff
@@ -291,7 +293,7 @@ class ParallelMatrixOperations {
       // distribute blocks of summarizedBlocks (result of multiplication) to corresponding slaves
       result = summarizedBlocks.block(0, 0, offsets[1], r);
 
-      for (int rankSlave : utils::MasterSlave::allSlaves()) {
+      for (Rank rankSlave : utils::MasterSlave::allSlaves()) {
         int off       = offsets[rankSlave];
         int send_rows = offsets[rankSlave + 1] - offsets[rankSlave];
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/acceleration/impl/ParallelMatrixOperations.hpp:298]: Offsets are rank related, but they aren't ranks. So this is where things get a bit tricky.
We could add RankOffset as an alias though.

File: src/precice/types.hpp (hunk 2 above)
Comment (fsimonis) [reply src/precice/types.hpp:23]: Used in nearest projection and could really go into the query package.

A nearest projection can match vertices, edge and triangles, so the type of the match is unclear.

**Pull request merged**